Show only the current user's payment methods when editing subscriptions

RadoslavGeorgiev opened the pull request:
Fixes WOOPMNT-5508

#### Changes proposed in this Pull Request

When editing subscriptions in WooCommerce admin, the payment method dropdown was showing all customers' saved payment methods instead of only the selected customer's methods for new subscriptions. For existing subscriptions, payment methods were not reloaded when the customer was changed. This PR fixes the issue by dynamically loading payment methods when the customer is changed.

**Key changes:**
- Replaced static JavaScript with a React/TypeScript component that fetches payment methods via AJAX when the customer changes
- Added `UserTokenCache` class to cache fetched tokens and manage loading/error states
- Added AJAX endpoint (`wcpay_get_user_payment_tokens`) to fetch payment tokens for a specific user
- Shows appropriate UI states: loading indicator, error messages, and "Please select a customer first" placeholder
- Maintains backward compatibility with older WooCommerce Subscriptions versions

**How can this code break?**
- AJAX request could fail if nonce expires or network issues occur → handled with error state display
- Customer select uses select2 which doesn't emit standard events → handled by listening to both select2 and native change events

#### Testing instructions

1. Create a store with multiple customers who have saved payment methods.
1. Create a subscription with one of the customers.
1. Go to WooCommerce → Subscriptions → Edit the subscription.
1. Verify payment methods dropdown shows only the current customer's saved payment methods.
1. Change the customer using the customer selector dropdown.
1. Verify a loading indicator appears briefly.
1. Verify the new customer's payment methods are loaded and displayed.
1. Test with a customer who has no saved payment methods - should show "Please select a payment method" placeholder
1. Test with no customer selected - should show "Please select a customer first"

Note: You will see failing workflows, but they should not be related to the changes from this PR.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times.
- [x] Covered with tests (or have a good reason not to test in description ☝️)

RadoslavGeorgiev committed:
Rather than displaying all saved payment methods, indicate that customer is needed.

RadoslavGeorgiev committed:
Initial changes to make the payment methods dynamic.

RadoslavGeorgiev committed:
Order the JS component

RadoslavGeorgiev committed:
Move the script to TSX

RadoslavGeorgiev committed:
Add types to the mix

RadoslavGeorgiev committed:
Add loading states, switch to a span to preserve HTML

RadoslavGeorgiev committed:
Initial round of tests

RadoslavGeorgiev committed:
Extract the cache into separate class, separate rendering and updates

RadoslavGeorgiev committed:
Updated comments

RadoslavGeorgiev committed:
Changelog entry

RadoslavGeorgiev committed:
Undo an accidental change

RadoslavGeorgiev committed:
Merge branch 'develop' into woopmnt-5508-shows-all-customers-saved-payment-methods-instead-of

RadoslavGeorgiev committed:
Unit tests

RadoslavGeorgiev committed:
Fixing AJAX calls for unit tests

RadoslavGeorgiev committed:
Merge branch 'develop' into woopmnt-5508-shows-all-customers-saved-payment-methods-instead-of

RadoslavGeorgiev committed:
Merge branch 'develop' into woopmnt-5508-shows-all-customers-saved-payment-methods-instead-of

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,280 @@
+		userId = newUserId;
+		render();
+
+		// Looaded, loading, or errored out, we do not need to load anything.
```
```suggestion
		// Loaded, loading, or errored out, we do not need to load anything.
```

RadoslavGeorgiev replied:
Typo fixed in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/af31e44dc5a3555cd4b142860c6633520c9e7579

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,280 @@
+	const internalCallback = () =>
+		callback( parseInt( customerUserSelect.value, 10 ) || 0 );
+
+	// Add the listner with the right technique, as select2 does not emit <select> events.
```
```suggestion
	// Add the listener with the right technique, as select2 does not emit <select> events.
```

RadoslavGeorgiev replied:
Typo fixed in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/af31e44dc5a3555cd4b142860c6633520c9e7579

frosso commented on client/subscription-edit-page/types.d.ts:
```diff
@@ -0,0 +1,63 @@
+
+/**
+ * Token represents a payment method token for a user
+ */
+export interface Token {
+	tokenId: number;
+	displayName: string;
+}
```
(minor) the Token type should also have a `isDefault` boolean, since I see it in `render_custom_payment_meta_input`. Although it doesn't seem to be used by the JS.

RadoslavGeorgiev replied:
I added the property in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/4043b4914c6f4e20531c154d2396f124212d7fe1 and started using it in 8fbab96bccb8871330f8e5be63670b09236e456a. The latter is not perfect yet, as it simply done in `render()`.

RadoslavGeorgiev replied:
Now it's an effect in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/3275371ee267adeda5a1080287ffc53c23e94411.

frosso commented on client/subscription-edit-page/types.d.ts:
```diff
@@ -0,0 +1,63 @@
+ */
+export interface PaymentMethodElementProps {
+	element: HTMLSpanElement;
+}
```
```suggestion
```

This doesn't seem to be used anywhere

RadoslavGeorgiev replied:
It was just a left-over. Removed in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/4043b4914c6f4e20531c154d2396f124212d7fe1

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,280 @@
+): ( () => void ) => {
+	const customerUserSelect = document.getElementById(
+		'customer_user'
+	) as HTMLSelectElement | null;
```
(minor) rather than using `as`, we can check the type returned by the `getElementById`. This way we don't need to force the casting of the type.

```suggestion
	const element = document.getElementById( 'customer_user' );
	const customerUserSelect =
		element instanceof HTMLSelectElement ? element : null;
```

RadoslavGeorgiev replied:
Addressed in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/4043b4914c6f4e20531c154d2396f124212d7fe1

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,280 @@
+	const input = element.querySelector( 'select,input' ) as
+		| HTMLSelectElement
+		| HTMLInputElement
+		| null;
```
(minor) similar to above - we can check the returned type, instead of forcing it.

```suggestion
	const input = element.querySelector( 'select,input' );
	if ( ! input ) {
		return;
	}
	if (
		! (
			input instanceof HTMLSelectElement ||
			input instanceof HTMLInputElement
		)
	) {
		return;
	}
```

RadoslavGeorgiev replied:
Addressed in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/4043b4914c6f4e20531c154d2396f124212d7fe1

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,280 @@
+const addPaymentMethodDropdowns = (): void => {
+	// Use a centralized cache for user tokens.
+	const cache = new UserTokenCache();
+
+	// There should be a single element on the page, but still make sure to iterate over all of them.
+	document
+		.querySelectorAll( '.wcpay-subscription-payment-method' )
+		.forEach( ( element ) => {
+			setupPaymentSelector( element as HTMLSpanElement, cache );
+		} );
```
(minor) to remove the forced casting 

```suggestion
	document
		.querySelectorAll< HTMLSpanElement >(
			'.wcpay-subscription-payment-method'
		)
		.forEach( ( element ) => {
			setupPaymentSelector( element, cache );
		} );
```

RadoslavGeorgiev replied:
Addressed in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/4043b4914c6f4e20531c154d2396f124212d7fe1.

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,280 @@
+
+	if ( ! customerUserSelect ) {
+		return (): void => {
+			// No-op cleanup function when element is not found
```
( ⛏️ )
```suggestion
			// No-op cleanup function when an element is not found
```

RadoslavGeorgiev replied:
Typo fixed in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/af31e44dc5a3555cd4b142860c6633520c9e7579

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,280 @@
+) => {
+	on: ( event: string, handler: () => void ) => void;
+	off: ( event: string, handler: () => void ) => void;
+};
```
Is this needed? I don't see TS blowing up if it gets removed 🤷

RadoslavGeorgiev replied:
Removed in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/4043b4914c6f4e20531c154d2396f124212d7fe1. I thought it would be good to get some types there, but this is more than redundant.

frosso replied:
It looks like it came back :D

RadoslavGeorgiev replied:
Claude seems to be stubborn, and this one went under my radar... Removed again in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/59b6b22bd99e33d194327ea7455cbb583609dcf7, everything seems to still work well.

RadoslavGeorgiev replied:
And it's back again! 🙈 

The front-end works fine, but tests are always failing without the declaration. I can't figure out a way around it.

frosso reviewed (commented):
The pub/sub pattern on the token cache is a really cool concept; however I don't think it's necessary.
There is a lot of "manual" rendering happening, despite the implementation of `PaymentMethodSelect` as a ReactJS component.
The current approach is a bit "mixed", and seems to defeat the purpose of using ReactJS. React should handle re-rendering automatically through state changes, instead of manual function calls.
For example, I see that the `onChange` prop is drilled down to the `select`, but at every change it manually re-renders the whole root. It's _kinda_ like ReactJS works, but re-invented :) 

Here's a possible alternative.
Make the `cache` a global, it can just be a key-value-pair of userId and their tokens:
```
const cache: Record< number, Token[] > = {};
```

On the `setupPaymentSelector`, you can substitute most of the rendering/listening to just creating the root and rendering the `PaymentMethodManager` - which is a helper component around the `PaymentMethodSelect`:
```
const setupPaymentSelector = ( element: HTMLSpanElement ): void => {
	const data = JSON.parse(
		element.getAttribute( 'data-wcpay-pm-selector' ) || '{}'
	) as WCPayPMSelectorData;

	const initialUserId = data.userId ?? 0;
	const initialValue = data.value ?? 0;

	// Initial population of cache with pre-loaded tokens
	if ( initialUserId && data.tokens ) {
		cache[ initialUserId ] = data.tokens;
	}

	// In older Subscriptions versions, there was just a simple input.
	const input = element.querySelector( 'select,input' ) as
		| HTMLSelectElement
		| HTMLInputElement
		| null;
	if ( ! input ) {
		return;
	}

	const root = createRoot( element );
	root.render(
		<PaymentMethodManager
			inputName={ input.name }
			initialUserId={ initialUserId }
			initialValue={ initialValue }
			config={ data }
		/>
	);
};
```

Then, the `PaymentMethodManager` can take care of listening to DOM changes on the `customer_user` input and fetching data, providing the data to the `select`:
```
const PaymentMethodManager = ( {
	inputName,
	initialUserId,
	initialValue,
	config,
}: {
	inputName: string;
	initialUserId: number;
	initialValue: number;
	config: WCPayPMSelectorData;
} ) => {
	const [ userId, setUserId ] = useState< number >( initialUserId );
	const [ selectedToken, setSelectedToken ] = useState< number >(
		initialValue
	);
	const [ isLoading, setIsLoading ] = useState< boolean >( false );
	const [ loadingError, setLoadingError ] = useState< string >( '' );

	// Handle customer select changes
	useEffect( () => {
		const updateSelectedToken = ( tokens: Token[] ) => {
			const selectedTokenId = tokens.find(
				( token ) => token.tokenId === selectedToken
			)?.tokenId;
			setSelectedToken( selectedTokenId || 0 );
		};
		return addCustomerSelectListener( async ( newUserId ) => {
			setUserId( newUserId );

			// Check if already in cache
			const tokens = cache[ newUserId ];

			// If already loaded, no need to fetch again.
			if ( tokens ) {
				updateSelectedToken( tokens );
				return;
			}

			setIsLoading( true );

			try {
				const response = await fetchUserTokens(
					newUserId,
					config.ajaxUrl,
					config.nonce
				);
				if ( undefined === response ) {
					throw new Error(
						__(
							'Failed to fetch user tokens. Please reload the page and try again.',
							'woocommerce-payments'
						)
					);
				}

				cache[ newUserId ] = response.tokens;
				updateSelectedToken( response.tokens );
				setLoadingError( '' );
			} catch ( error ) {
				setLoadingError(
					error instanceof Error
						? error.message
						: __( 'Unknown error', 'woocommerce-payments' )
				);
			}

			setIsLoading( false );
		} );
	}, [ config.ajaxUrl, config.nonce, selectedToken ] );

	if ( loadingError ) {
		return <strong>{ loadingError }</strong>;
	}

	if ( isLoading || ! cache[ userId ] ) {
		return <>{ __( 'Loading…', 'woocommerce-payments' ) }</>;
	}

	return (
		<select name={ inputName } defaultValue={ initialValue } key={ userId }>
			<option value={ 0 } key="select" disabled>
				{ __(
					'Please select a payment method',
					'woocommerce-payments'
				) }
			</option>
			{ cache[ userId ].map( ( token ) => (
				<option value={ token.tokenId } key={ token.tokenId }>
					{ token.displayName }
				</option>
			) ) }
		</select>
	);
};
```

What do you think?

RadoslavGeorgiev committed:
Fix typos

RadoslavGeorgiev committed:
Remove unused type, add the default token property, remove jQuery's definition, don't force element types.

RadoslavGeorgiev committed:
Use the default token when changing users

RadoslavGeorgiev committed:
Simplify the select

RadoslavGeorgiev committed:
pdate subscription-edit-page tests for refactored cache functions

  - Convert UserTokenCache class tests to pure function tests (startLoading,
    tokensLoaded, loadingFailed, hasEntry, getUserEntry, userHasToken,
    getDefaultTokenId)
  - Update PaymentMethodSelect tests to use new props (initialValue,
    initialUserId, initialCache, nonce, ajaxUrl)
  - Add immutability tests for cache functions
  - Add jQuery type declaration to index.tsx for TypeScript compilation
  - Adjust tests for new placeholder behavior (only shown when value === 0)

RadoslavGeorgiev committed:
Merge branch 'develop' into woopmnt-5508-shows-all-customers-saved-payment-methods-instead-of

RadoslavGeorgiev committed:
Export fetchUserTokens and addCustomerSelectListener, add tests for them.

RadoslavGeorgiev committed:
Merge branch 'woopmnt-5508-shows-all-customers-saved-payment-methods-instead-of' of github.com:Automattic/woocommerce-payments into woopmnt-5508-shows-all-customers-saved-payment-methods-instead-of

RadoslavGeorgiev commented:
@frosso I did go a bit overboard with the cache as a separate class. I went back and combined some of your suggestions with a bit of cleaner logic.

Previously, I had things in a state that was a tad too messy, and it was difficult for me to break things down into testable and graspable methods. The `PaymentMethodSelect` was just too long to be stable. Now I switched to local cache, simplified things a bit, and I think that the separation of methods is satisfying. WDYT?

> For example, I see that the onChange prop is drilled down to the select

Actually, that was due to me not knowing about the `defaultValue` prop and linters driving me nuts. Your suggestion helped me see that! 🙇

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,285 @@
+	const entry = getUserEntry( cache, userId );
+	if ( undefined === entry || entry.loading ) {
+		return <>{ __( 'Loading…', 'woocommerce-payments' ) }</>;
+	} else if ( entry.loadingError ) {
```
```suggestion
	}
	if ( entry.loadingError ) {
```

( ⛏️ ) since there's an early return on the block above, the `else` doesn't seem to be needed

RadoslavGeorgiev replied:
Separated in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/7257a80c0d4beac8a5e1419e35663f6e56d4539a.

frosso committed:
Merge branch 'develop' into woopmnt-5508-shows-all-customers-saved-payment-methods-instead-of

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,285 @@
+ * This could be a shorter method, but because the customer select
+ * element uses select2, it does not emit the typical `change` event.
+ *
+ * @param {(userId: number) => void} callback The callback to call when the customer is changed.
```
The JS Docs need to be updated to reflect the new arguments

RadoslavGeorgiev replied:
Done in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/7257a80c0d4beac8a5e1419e35663f6e56d4539a.

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,285 @@
+				setValue( value );
+			}
+		}
+	}, [ userId, value, cache ] );
```
```suggestion
	// If there is no value but a user, try to fall back to the default token.
	useEffect( () => {
		if ( 0 !== userId && value === 0 ) {
			const defaultTokenId = getDefaultTokenId( cache, userId );
			if ( value !== defaultTokenId ) {
				setValue( defaultTokenId );
			}
		}
	}, [ userId, value, cache ] );
```

Given the description above the `useEffect`, should this have been the intent?


Also, is `cache` necessary as part of the dependencies?

RadoslavGeorgiev replied:
> Given the description above the `useEffect`, should this have been the intent?

Absolutely. I'm not sure how I let this through. Maybe there was some TS error in the test file, and I was actually looking at an outdated front-end, as this was one of the later changes. 🤔

I updated it to use `defaultTokenId`.

> Also, is cache necessary as part of the dependencies?

Once tokens finish loading for a new user, we need to use the cache update to trigger the effect. `userId` and `value` are not directly changed when tokens get loaded.

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,285 @@
+	if (
+		! input ||
+		! ( input instanceof HTMLSelectElement ) ||
+		! ( input instanceof HTMLInputElement )
```
`input` can never be both a `HTMLSelectElement` and a `HTMLInputElement`.
In my initial suggestion, I had separate the two conditions:
```js
	if ( ! input ) {
		return;
	}
	if (
		! (
			input instanceof HTMLSelectElement ||
			input instanceof HTMLInputElement
		)
	) {
		return;
	}
```


If you wanted all the conditions into one check, consider something like this:
```js
	if ( ! input &&
		! (
			input instanceof HTMLSelectElement ||
			input instanceof HTMLInputElement
		)
	) {
		return;
	}
```

RadoslavGeorgiev replied:
🤦 

Rookie mistake, combined with looking at an outdated compiled scripts, as I mentioned in [the other comment](https://github.com/Automattic/woocommerce-payments/pull/11143#discussion_r2580320538). Fixed in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/7257a80c0d4beac8a5e1419e35663f6e56d4539a

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,285 @@
+			setUserId,
+			setCache
+		);
+	}, [ userId, cache, ajaxUrl, nonce ] );
```
The `userId` isn't part of the dependencies of `addCustomerSelectListener` - is it necessary to destroy and re-construct the observer each time the user id changes?

RadoslavGeorgiev replied:
It was a dependency before I moved some logic to `addCustomerSelectListener`. Removed in https://github.com/Automattic/woocommerce-payments/commit/7257a80c0d4beac8a5e1419e35663f6e56d4539a

frosso commented on client/subscription-edit-page/user-token-cache.ts:
```diff
@@ -0,0 +1,111 @@
+	cachedData: CachedUserData,
+	userId: number
+): CachedUserDataItem | undefined => {
+	return cachedData.find( ( userData ) => userData.userId === userId );
```
Have you considered implementing the cache as a key-value-pair (where the key is the user id) for faster `O(1)` access? It seems that we're always looking for an entry by its user id.

RadoslavGeorgiev replied:
Yes, that's a good idea. Done in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/59b6b22bd99e33d194327ea7455cbb583609dcf7.

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,285 @@
+	const value = data.value ?? 0;
+
+	// Initial population.
+	const cache: CachedUserData = [];
```
It looks like each instance of a `.wcpay-subscription-payment-method` is getting its own cache. And this array is also re-created as a `useState` down in the `PaymentMethodSelect`.
What do you think about having the cache as a global key-value-pair? Or it can also be a global variable in the `client/subscription-edit-page/user-token-cache.ts` file, so you don't need to pass it as an argument each time you want to call one of the utility functions. And you don't need to prop-drill it.

RadoslavGeorgiev replied:
~At this stage, I find it arbitrary. I decided to go away from the global cache (it was never used in more than one instance anyway), and managing the cache in state made it relatively easier to break things down.~

~If I move to a global cache as an object to avoid having to pass it as an argument, I'll still need to contain loading and failure state management within the select. I feel like that brings me back to a bigger and unwieldier method, and I don't see a significant upside anyway.~

~I'll try to see how it works, and how much/what will need to still be passed around, but I'm not promising anything yet.~

Nevermind, done. It's so much cleaner. Thank you for being persisent here!

frosso reviewed (commented):
It doesn't look like it's functional anymore - I think because of the condition on the select?

RadoslavGeorgiev committed:
A bunch of minor changes from the latest feedback.

RadoslavGeorgiev committed:
Use userIds as cache keys

RadoslavGeorgiev committed:
Re-remove the jQuery declaration.

RadoslavGeorgiev committed:
Simplify cache

RadoslavGeorgiev committed:
Move error handling to fetchUserTokens

RadoslavGeorgiev committed:
Rather than using a separate effect, set the default token upon load.

RadoslavGeorgiev committed:
Updated tests

RadoslavGeorgiev committed:
Use the default token before AND after loading

frosso committed:
Merge branch 'develop' into woopmnt-5508-shows-all-customers-saved-payment-methods-instead-of

frosso commented on client/subscription-edit-page/__tests__/index.test.tsx:
```diff
@@ -0,0 +1,545 @@
+	PaymentMethodSelect,
+	fetchUserTokens,
+	clearTokenCache,
+} from '../index';
```
```suggestion
} from '..';
```

(minor) since `index` is not needed

RadoslavGeorgiev replied:
removed the `/index` in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/d3ff70979cd1e024207de9a953b5e654e0a1b771

frosso commented on client/subscription-edit-page/index.tsx:
```diff
@@ -0,0 +1,279 @@
+				);
+			}
+		} );
+	}, [ value, ajaxUrl, nonce ] );
```
Is `value` needed as part of the dependencies? i.e.: if the selected token is updated, do we need to destroy and recreate the listener on the "Recipient" input?

RadoslavGeorgiev replied:
Again, this is a left-over. I removed it in https://github.com/Automattic/woocommerce-payments/pull/11143/commits/d3ff70979cd1e024207de9a953b5e654e0a1b771. I need to get used to monitoring dependencies.

frosso replied:
Maybe we could add a linter to ensure all the dependencies needed are provided, and no extraneous ones are added. If they do need to be excluded (or added), a `eslint-ignore` can be added.

We do that on WooPay with `react-hooks/exhaustive-deps`, for example 🤷

RadoslavGeorgiev replied:
> We do that on WooPay with `react-hooks/exhaustive-deps`, for example 🤷

The package is in place here as well. That's why I was adding too many dependencies early, and not realizing it.

frosso reviewed (approved):
Nice, thank you for entertaining my comments! No major issues from me :) I re-tested and it works as expected

RadoslavGeorgiev committed:
Remove `value` as a dependency of the customer select listener; Fix a redundand `/index`.
